k8s/jenkins/docs/CONFIGURE_K8S_PLUGIN_IN_JENKINS.md

```diff
@@ -0,0 +1,196 @@
+
+![Architecture diagram](resources/jenkins-k8s-app-architecture-with-k8s-plugin.png)
+
+
+
+# Deploying Agents in K8s
+
+## Jenkins service account
+
+1. Create a Jenkins service account to integrate Jenkins to Kubernetes to deploy agents: 
+
+```
+cat > jenkins-service-account.yaml << EOF
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: $APP_INSTANCE_NAME-serviceaccount
+EOF
+kubectl apply -f jenkins-service-account.yaml -n $NAMESPACE
+```
+
+2. Create a service role with permissions to deploy agent pods:
+```
+cat > jenkins-service-role.yaml << EOF
+---
+kind: Role
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: $APP_INSTANCE_NAME-role
+rules:
+- apiGroups: [""]
+  resources: ["pods"]
+  verbs: ["create","delete","get","list","patch","update","watch"]
+- apiGroups: [""]
+  resources: ["pods/exec"]
+  verbs: ["create","delete","get","list","patch","update","watch"]
+- apiGroups: [""]
+  resources: ["pods/log"]
+  verbs: ["get","list","watch"]
+- apiGroups: [""]
+  resources: ["events"]
+  verbs: ["watch"]
+- apiGroups: [""]
+  resources: ["secrets"]
+  verbs: ["get"]
+EOF
+kubectl apply -f jenkins-service-role.yaml -n $NAMESPACE
+```
+
+3. Create a role binding for Jenkins service account:
+```
+cat > jenkins-service-role-binding.yaml << EOF
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: $APP_INSTANCE_NAME-rolebinding
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: $APP_INSTANCE_NAME-role
+subjects:
+- kind: ServiceAccount
+  name: $APP_INSTANCE_NAME-serviceaccount
+EOF
+kubectl apply -f jenkins-service-role-binding.yaml -n $NAMESPACE
+```
+
+## Configure Kubernetes plugin
+
+1. Install [Jenkins Kubernetes Plugin](https://plugins.jenkins.io/kubernetes/).
+
+2. Add the service account token to Jenkins, first get the token using the next command:
+```
+# Service Account Token
+kubectl get secret $(kubectl get sa $APP_INSTANCE_NAME-serviceaccount \
+-n $NAMESPACE -o jsonpath="{.secrets[0].name}") \
+-n $NAMESPACE -o jsonpath="{.data.token}" | base64 --decode
+```
+3. Create a secret text in Jenkins with the token:
+
+![Service account token](resources/configure-jenkins-k8s-plugin1.png)
+
+4. Get the CA certificate and add it to the plugin configuration:
+```
+# CA Certificate
+kubectl get secret $(kubectl get sa $APP_INSTANCE_NAME-serviceaccount \
+-n $NAMESPACE -o jsonpath="{.secrets[0].name}") \
+-n $NAMESPACE -o jsonpath={.data.'ca\.crt'} | base64 --decode
+```
+![Add CA certificate](resources/configure-jenkins-k8s-plugin2.png)
+
+5. Get the jenkins service pod ip address and add it to the plugin configuration as Jenkins Url.
+```
+echo http://$(kubectl get svc $APP_INSTANCE_NAME-jenkins-ui -n $NAMESPACE -o jsonpath="{.spec.clusterIP}"):8080
+```
+
+6. Since the agent connector service is exposed as ClusterIP is required to use the option Jenkins tunnel in the plugin, you can get the internal service name with the next command:
```

Comment:
> Since the agent connector service is exposed as ClusterIP is required to use the option Jenkins tunnel in the plugin,

Because you use ClusterIP to configure the Jenkins URL, you expose the agent connector service through ClusterIP. You can use the following command to get the internal service name: